src/backend/ctrl_flow_struct/condition.rs

```diff
@@ -0,0 +1,171 @@
+use std::fmt;
+use std::ptr;
+use typed_arena::Arena;
+
+pub struct BaseCondition<'cd, T: 'cd>(&'cd BaseConditionS<'cd, T>);
```

Comment:
Not going to repeat this for all the structs introduced, add doc comments explaining what the struct is for, what API does it expose to its users.

Additionally, for public functions to be called by the structs users include what pre-conditions must hold on its arguments and what the  expected result is.

As normal comments include: if there are some tricky implementation details, thing that aren't obvious and likely to break if someone else makes some changes to it.

Reply:
I think that this module is going to need to change a lot in order to implement refinements (i.e. `if (a) doA(); if (!a) doB();` ==> `if (a) doA(); else doB();`) so I'm going to hold off on documenting all the APIs of this module until those changes have happened.